web/turkey/survey/views.py

```diff
@@ -1,106 +1,215 @@
+
+    def get_associated_models(self, map, task_interaction):
+        """
+        Get models of steps or auditors that apply to the task
+        linked to by task_interaction
+        """
+        associated_models = []
+
+        for model_name in map.values():
+            model = apps.get_model('survey', model_name)
+            associated_models.extend(model.objects.filter(task=task_interaction.task))
+
+        return associated_models
+
+    def associated_models_have_data(self, map, task_interaction):
+        """
+        Find out if the models associated with a task_interaction
+        already have data instances. This serves to show us whether there has
+        already been a submission or not
+        """
+        associated_models = self.get_associated_models(map, task_interaction)
+        for model in associated_models:
+            if model.has_data_for_task_interaction(task_interaction):
+                return True
+
+        return False
 
     def post(self, request, **kwargs):
-        submission = request.data
-        # TODO: What to do when it's not found?
+        with transaction.atomic():
+            task_interaction_model = self.get_task_interaction(kwargs['pk'])
+            if not task_interaction_model:
+                return Response(status=status.HTTP_400_BAD_REQUEST)
+            try:
+                self.process_submission(request.data, task_interaction_model)
+                return Response(status=status.HTTP_201_CREATED)
+            except ValidationError:
+                return Response(status=status.HTTP_400_BAD_REQUEST)
 
-        try:
-            task_interaction_model = TaskInteraction.objects \
-                .get(pk=int(kwargs['pk']))
-        except TaskInteraction.DoesNotExist:
-            return Response(status=status.HTTP_400_BAD_REQUEST)
-        if task_interaction_model.completed:
+
+class StepSubmission(RecordSubmission):
+    def process_submission(self, submission_data, task_interaction):
+        if task_interaction.task.external:
+            raise ValidationError(_('Submitting steps on an internal HIT'))
+
+        if self.associated_models_have_data(NAME_TO_STEP, task_interaction):
+            raise ValidationError(_('Step data already submitted for this task'))
+
+        self.save_data_to_mapped_models(submission_data['steps'],
+                                        NAME_TO_STEP,
+                                        task_interaction)
+
+
+class AuditorSubmission(RecordSubmission):
+    def process_submission(self, submission_data, task_interaction):
+        if self.associated_models_have_data(NAME_TO_AUDITOR, task_interaction):
+            raise ValidationError(_('Auditor data already submitted for this task'))
+
+        self.save_data_to_mapped_models(submission_data['auditors'],
+                                        NAME_TO_AUDITOR,
+                                        task_interaction)
+
+
+class CreateTaskInteractionView(APIView):
```

Comment:
If the task is external, we need AuditorHandler to be able to create a task interaction and know what it's primary key is so it can submit its data accordingly